```
	 FAIL	github.com/cilium/cilium/pkg/mountinfo [build failed]
	 FAIL
	 Makefile:148: recipe for target 'tests-privileged' failed
	 make: Leaving directory '/home/vagrant/go/src/github.com/cilium/cilium'
	 
Stderr:
 	 # github.com/cilium/cilium/pkg/mountinfo [github.com/cilium/cilium/pkg/mountinfo.test]
	 pkg/mountinfo/mountinfo_privileged_test.go:30:10: undefined: MountInfoTestSuite
	 make: *** [tests-privileged] Error 1
```
from https://jenkins.cilium.io/job/Cilium-PR-Runtime-4.9/1009/testReport/junit/(root)/Suite-runtime/RuntimePrivilegedUnitTests_Run_Tests/

--- a/pkg/bpf/bpffs_linux.go
+++ b/pkg/bpf/bpffs_linux.go
@@ -72,7 +72,7 @@ func mapPathFromMountInfo(name string) string {
 		}
 
 		for _, mountInfo := range mountInfos {
-			if mountInfo.FilesystemType == mountinfo.FilesystemTypeBPFFS {
+			if mountInfo.FilesystemType == "bpf" {
 				mountInfoPrefix = filepath.Join(mountInfo.MountPoint, mapPrefix)
 				return
 			}

--- a/pkg/cgroups/cgroups.go
+++ b/pkg/cgroups/cgroups.go
@@ -62,7 +62,7 @@ func mountCgroup() error {
 		return fmt.Errorf("%s is a file which is not a directory", cgroupRoot)
 	}
 
-	if err := unix.Mount("none", cgroupRoot, mountinfo.FilesystemTypeCgroup2, 0, ""); err != nil {
+	if err := unix.Mount("none", cgroupRoot, "cgroup2", 0, ""); err != nil {
 		return fmt.Errorf("failed to mount %s: %s", cgroupRoot, err)
 	}
 

--- a/pkg/mountinfo/mountinfo.go
+++ b/pkg/mountinfo/mountinfo.go
@@ -67,8 +67,8 @@ func parseMountInfoFile(r io.Reader) ([]*MountInfo, error) {
 		}
 
 		// Extract fields from both sides of mountinfo
-		mountInfoLeft := strings.Split(strings.TrimSpace(mountInfoSeparated[0]), " ")
-		mountInfoRight := strings.Split(strings.TrimSpace(mountInfoSeparated[1]), " ")
+		mountInfoLeft := strings.Split(mountInfoSeparated[0], " ")
+		mountInfoRight := strings.Split(mountInfoSeparated[1], " ")
 
 		// Before '-' separator there should be 6 fields and unknown
 		// number of optional fields

--- a/pkg/mountinfo/mountinfo.go
+++ b/pkg/mountinfo/mountinfo.go
@@ -19,14 +19,18 @@ import (
 	"fmt"
 	"io"
 	"os"
+	"path/filepath"
 	"strconv"
 	"strings"
+
+	"golang.org/x/sys/unix"
 )
 
 const (
-	// FilesystemType names for filesystem which are used in /proc/pid/mountinfo
-	FilesystemTypeBPFFS   = "bpf"
-	FilesystemTypeCgroup2 = "cgroup2"
+	// FilesystemType superblock magic numbers for filesystems,
+	// to be used for IsMountFS.
+	FilesystemTypeBPFFS   = unix.BPF_FS_MAGIC
+	FilesystemTypeCgroup2 = unix.CGROUP2_SUPER_MAGIC
 
 	mountInfoFilepath = "/proc/self/mountinfo"
 )
@@ -129,35 +133,41 @@ func GetMountInfo() ([]*MountInfo, error) {
 	return parseMountInfoFile(fMounts)
 }
 
-func isMountFS(mountInfos []*MountInfo, mntType string, mapRoot string) (bool, bool) {
-	var mapRootMountInfo *MountInfo
+// IsMountFS returns two boolean values, checking
+//  - whether the path is a mount point;
+//  - if yes, whether its filesystem type is mntType.
+//
+// Note that this function can not detect bind mounts,
+// and is not working properly when path="/".
+func IsMountFS(mntType int64, path string) (bool, bool, error) {
+	var st, pst unix.Stat_t
 
-	for _, mountInfo := range mountInfos {
-		if mountInfo.MountPoint == mapRoot {
-			mapRootMountInfo = mountInfo
-			break
-		}
-	}
-
-	if mapRootMountInfo == nil {
-		return false, false
-	}
-
-	if mapRootMountInfo.FilesystemType == mntType {
-		return true, true
-	}
-	return true, false
-}
-
-// IsMountFS returns two boolean values:checks whether the current mapRoot:
-// - whether the current mapRoot has any mount
-// - whether that mount's filesystem is of type mntType
-func IsMountFS(mntType string, mapRoot string) (bool, bool, error) {
-	mountInfos, err := GetMountInfo()
+	err := unix.Lstat(path, &st)
 	if err != nil {
-		return false, false, err
+		if err == unix.ENOENT {
+			// non-existent path can't be a mount point
+			return false, false, nil
+		}
+		return false, false, &os.PathError{Op: "lstat", Path: path, Err: err}
 	}
 
-	mounted, mntTypeInstance := isMountFS(mountInfos, mntType, mapRoot)
-	return mounted, mntTypeInstance, nil
+	parent := filepath.Dir(path)
+	err = unix.Lstat(parent, &pst)
+	if err != nil {
+		return false, false, &os.PathError{Op: "lstat", Path: parent, Err: err}
+	}
+	if st.Dev == pst.Dev {
+		// parent has the same dev -- not a mount point
+		return false, false, nil
+	}
+
+	// Check the fstype
+	fst := unix.Statfs_t{}
+	err = unix.Statfs(path, &fst)
+	if err != nil {
+		return true, false, &os.PathError{Op: "statfs", Path: path, Err: err}
+	}
+
+	return true, fst.Type == mntType, nil
+
 }

--- a/pkg/mountinfo/mountinfo_privileged_test.go
+++ b/pkg/mountinfo/mountinfo_privileged_test.go
@@ -1,0 +1,59 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build privileged_tests
+
+package mountinfo
+
+import (
+	"io/ioutil"
+	"os"
+
+	"golang.org/x/sys/unix"
+
+	. "gopkg.in/check.v1"
+)
+
+type MountInfoPrivilegedTestSuite struct{}
+
+var _ = Suite(&MountInfoPrivilegedTestSuite{})
+
+// TestIsMountFSbyMount tests the public function IsMountFS by performing
+// an actual mount.
+func (s *MountInfoPrivilegedTestSuite) TestIsMountFSbyMount(c *C) {
+	tmpDir, err := ioutil.TempDir("", "IsMountFS_")
+	c.Assert(err, IsNil)
+	defer os.RemoveAll(tmpDir)
+
+	mounted, matched, err := IsMountFS(unix.TMPFS_MAGIC, tmpDir)
+	c.Assert(err, IsNil)
+	c.Assert(mounted, Equals, false)
+	c.Assert(matched, Equals, false)
+
+	err = unix.Mount("tmpfs", tmpDir, "tmpfs", 0, "")
+	c.Assert(err, IsNil)
+	defer unix.Unmount(tmpDir, unix.MNT_DETACH)
+
+	// deliberately check with wrong fstype
+	mounted, matched, err = IsMountFS(unix.PROC_SUPER_MAGIC, tmpDir)
+	c.Assert(err, IsNil)
+	c.Assert(mounted, Equals, true)
+	c.Assert(matched, Equals, false)
+
+	// now check with proper fstype
+	mounted, matched, err = IsMountFS(unix.TMPFS_MAGIC, tmpDir)
+	c.Assert(err, IsNil)
+	c.Assert(mounted, Equals, true)
+	c.Assert(matched, Equals, true)
+}

--- a/pkg/mountinfo/mountinfo_test.go
+++ b/pkg/mountinfo/mountinfo_test.go
@@ -21,6 +21,7 @@ import (
 	"testing"
 
 	"github.com/cilium/cilium/pkg/checker"
+	"golang.org/x/sys/unix"
 
 	. "gopkg.in/check.v1"
 )
@@ -600,39 +601,21 @@ func (s *MountInfoTestSuite) TestGetMountInfo(c *C) {
 	c.Assert(err, IsNil)
 }
 
-// TestIsMountFSPrivate tests the private function isMountFS which expects
-// mountinfo to be provided as a slice.
-func (s *MountInfoTestSuite) TestIsMountFSPrivate(c *C) {
-	r := bytes.NewBuffer([]byte(mountInfoContent))
-	mountInfos, err := parseMountInfoFile(r)
-	c.Assert(err, IsNil)
-
-	// /sys/fs/bpffs should be recognized as a BPFFS mount.
-	mounted, bpffsInstance := isMountFS(mountInfos, FilesystemTypeBPFFS, "/sys/fs/bpf")
-	c.Assert(mounted, Equals, true)
-	c.Assert(bpffsInstance, Equals, true)
-
-	// /sys/fs/cgroup/unified should be recognized as a cgroup2 mount.
-	mounted, cgroup2Instance := isMountFS(mountInfos, FilesystemTypeCgroup2, "/sys/fs/cgroup/unified")
-	c.Assert(mounted, Equals, true)
-	c.Assert(cgroup2Instance, Equals, true)
-
-	// /run/user/463 should be recognized as a mount, but not of BPFFS type.
-	mounted, bpffsInstance = isMountFS(mountInfos, FilesystemTypeBPFFS, "/run/user/463")
-	c.Assert(mounted, Equals, true)
-	c.Assert(bpffsInstance, Equals, false)
-
-	// /foo/bar shouldn't be found in mountinfo.
-	mounted, bpffsInstance = isMountFS(mountInfos, FilesystemTypeBPFFS, "/foo/bar")
-	c.Assert(mounted, Equals, false)
-	c.Assert(bpffsInstance, Equals, false)
-}
-
-// TestIsMountFS tests the public funcion IsMountFS which gets mountinfo from
-// /proc/self/mountinfo. We cannot expect every system and machine to have any
-// predictable mounts. The only thing which can be checked is whether there is
-// no error returned by mountinfo parser.
+// TestIsMountFS tests the public function IsMountFS. We cannot expect every
+// system and machine to have any predictable mounts, but let's try a couple
+// of very well known paths.
 func (s *MountInfoTestSuite) TestIsMountFS(c *C) {
-	_, _, err := IsMountFS(FilesystemTypeBPFFS, "/sys/fs/bpf")
+	mounted, matched, err := IsMountFS(unix.PROC_SUPER_MAGIC, "/proc")
 	c.Assert(err, IsNil)
+	c.Assert(mounted, Equals, true)
+	c.Assert(matched, Equals, true)
+
+	mounted, matched, err = IsMountFS(FilesystemTypeBPFFS, "/sys/fs/bpf")
+	c.Assert(err, IsNil)
+	// We can't expect /sys/fs/bpf is mounted, so only check fstype
+	// if it is mounted. IOW, if /sys/fs/bpf is a mount point,
+	// we expect it to be bpffs.
+	if mounted {
+		c.Assert(matched, Equals, true)
+	}
 }

